text/0195-kubernetes-resource-detection.md

```diff
@@ -0,0 +1,148 @@
+
+### Alternative: Using HOSTNAME for k8s.pod.name
+
+Many kubernetes detectors currently use `HOSTNAME` environment variable, which defaults to the Pod name. However, the `HOSTNAME` can be [modified in a few ways](https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#pod-s-hostname-and-subdomain-fields) in the pod spec. Kubernetes resource detectors may fall back to detecting the pod name using `HOSTNAME` if `K8S_POD_NAME` is not available, but this may cause user confusion in some cases.
+
+`HOSTNAME` is also [truncated to 64 characters](https://github.com/kubernetes/kubernetes/issues/4825) on some operating systems.
+
+### Alternative: Using Dependent Environment Variables
+
+Kubernetes supports defining environment variables based on [dependent environment variables](https://kubernetes.io/docs/tasks/inject-data-application/define-interdependent-environment-variables/#define-an-environment-dependent-variable-for-a-container). We can use this to transform the environment variables discovered through the downward API into the `OTEL_RESOURCE_ATTRIBUTES` environment variable:
+
+```yaml
+- name: OTEL_RESOURCE_ATTRIBUTES
+  value: k8s.pod.name=$(K8S_POD_NAME),k8s.pod.uid=$(K8S_POD_UID),k8s.namespace.name=$(K8S_NAMESPACE_NAME),k8s.node.name=$(K8S_NODE_NAME)
+```
+
+This approach has a few drawbacks:
+
+* `OTEL_RESOURCE_ATTRIBUTES` may need to include additional attibutes other than those used for kubernetes. This makes it harder to apply the same environment variables across all pods.
+* `OTEL_RESOURCE_ATTRIBUTES` doesn't have a way to specify the schema version it should apply to. This would conflict with the ability of [telemetry schemas](https://github.com/open-telemetry/oteps/blob/main/text/0152-telemetry-schemas.md#solution-summary) to convert from older versions of semantic conventions to newer ones.
+
+### Alternative: Support multiple OTEL_RESOURCE_ATTRIBUTES_* variables
+
+Proposed in [open-telemetry/opentelemetry-specification#2135](https://github.com/open-telemetry/opentelemetry-specification/issues/2135), SDKs would detect all environment variables starting with `OTEL_RESOURCE_ATTRIBUTES_`, and would add these to the detected resource attributes. The `OTEL_RESOURCE_ATTRIBUTES` prefix is trimmed, letters are lower-cased, and `_` are replaced with `.`.  For example, `OTEL_RESOURCE_ATTRIBUTES_K8S_NODE_NAME=foo` would become `k8s.node.name=foo`.
```

Comment:
This, in general, is ambiguous; take `cloud.availability_zone` defined [here](https://github.com/open-telemetry/opentelemetry-specification/blob/48ea76ccf06a027278e885e774b1371b91d6454d/specification/resource/semantic_conventions/cloud.md#cloud): env variable would presumably be `OTEL_RESOURCE_ATTRIBUTES_CLOUD_AVAILABILITY_ZONE` but the wording here would result in `cloud.availability.zone`. The main proposal doesn't suffer from this (the env variable <-> semantic convention is ad-hoc), but this alternative does suffer from it (e.g. would be a problem if we add a `_` on any of the relevant k8s semantic conventions)

Reply:
Thanks.  I added this to the list of drawbacks.